### PR TITLE
[feature-wip][array-type] Support more sub types.

### LINKED ISSUE #9465: [feature-wip][array-type] Support more sub types.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

This issue is a sub task of #7570 .

Support more sub types:
1. Array\<Boolean\>
2. Array\<TinyInt\>
3. Array\<SmallInt\>
4. Array\<Int\>
5. Array\<BigInt\>
6. Array\<LargeInt\>
7. Array\<Char\>
8. Array\<VarChar\>
9. Array\<String\>
10. Array\<Decimal\>
11. Array\<Date\>
12. Array\<DateTime\>

### Use case

```shell
# boolean
mysql> create table boolean_table (`key` int, value array<boolean>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into boolean_table values (0, null), (1, []), (2, [true, false]), (3, [null, false, null]);
mysql> select * from boolean_table;
+------+-----------------+
| key  | value           |
+------+-----------------+
|    1 | []              |
|    0 | NULL            |
|    3 | [NULL, 0, NULL] |
|    2 | [1, 0]          |
+------+-----------------+
4 rows in set (0.06 sec)

# tinyint
mysql> create table tinyint_table (`key` int, value array<tinyint>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into tinyint_table values (0, null), (1, []), (2, [1, 2]), (3, [null, 2, null]);
mysql> select * from tinyint_table;
+------+-----------------+
| key  | value           |
+------+-----------------+
|    1 | []              |
|    0 | NULL            |
|    2 | [1, 2]          |
|    3 | [NULL, 2, NULL] |
+------+-----------------+
4 rows in set (0.01 sec)

# smallint
mysql> create table smallint_table (`key` int, value array<smallint>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into smallint_table values (0, null), (1, []), (2, [1, 2]), (3, [null, 2, null]);
mysql> select * from smallint_table;
+------+-----------------+
| key  | value           |
+------+-----------------+
|    2 | [1, 2]          |
|    3 | [NULL, 2, NULL] |
|    0 | NULL            |
|    1 | []              |
+------+-----------------+
4 rows in set (0.01 sec)

# int
mysql> create table int_table (`key` int, value array<int>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into int_table values (0, null), (1, []), (2, [1, 2]), (3, [null, 2, null]);
mysql> select * from int_table;
+------+-----------------+
| key  | value           |
+------+-----------------+
|    3 | [NULL, 2, NULL] |
|    1 | []              |
|    0 | NULL            |
|    2 | [1, 2]          |
+------+-----------------+
4 rows in set (0.02 sec)

# bigint
mysql> create table bigint_table (`key` int, value array<bigint>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into bigint_table values (0, null), (1, []), (2, [1, 2]), (3, [null, 2, null]);
mysql> select * from bigint_table;
+------+-----------------+
| key  | value           |
+------+-----------------+
|    1 | []              |
|    0 | NULL            |
|    3 | [NULL, 2, NULL] |
|    2 | [1, 2]          |
+------+-----------------+
4 rows in set (0.02 sec)

# largeint
mysql> create table largeint_table (`key` int, value array<largeint>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into largeint_table values (0, null), (1, []), (2, [1, 2]), (3, [null, 2, null]);
mysql> select * from largeint_table;
+------+-----------------+
| key  | value           |
+------+-----------------+
|    1 | []              |
|    3 | [NULL, 2, NULL] |
|    2 | [1, 2]          |
|    0 | NULL            |
+------+-----------------+
4 rows in set (0.01 sec)

# float
mysql> create table float_table (`key` int, value array<float>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into float_table values (0, null), (1, []), (2, [1.5, 2.5]), (3, [null, 2.5, null]);
mysql> select * from float_table;
+------+-------------------+
| key  | value             |
+------+-------------------+
|    1 | []                |
|    2 | [1.5, 2.5]        |
|    3 | [NULL, 2.5, NULL] |
|    0 | NULL              |
+------+-------------------+
4 rows in set (0.02 sec)

# double
mysql> create table double_table (`key` int, value array<double>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into double_table values (0, null), (1, []), (2, [1.5, 2.5]), (3, [null, 2.5, null]);
mysql> select * from double_table;
+------+-------------------+
| key  | value             |
+------+-------------------+
|    1 | []                |
|    0 | NULL              |
|    3 | [NULL, 2.5, NULL] |
|    2 | [1.5, 2.5]        |
+------+-------------------+
4 rows in set (0.02 sec)

# char
mysql> create table char_table (`key` int, value array<char(10)>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into char_table values (0, null), (1, []), (2, ['1', '2']), (3, [null, '2', null]);
mysql> select * from char_table;
+------+-------------------+
| key  | value             |
+------+-------------------+
|    1 | []                |
|    0 | NULL              |
|    3 | [NULL, '2', NULL] |
|    2 | ['1', '2']        |
+------+-------------------+
4 rows in set (0.02 sec)

# varchar
mysql> create table varchar_table (`key` int, value array<varchar>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into varchar_table values (0, null), (1, []), (2, ['1', '2']), (3, [null, '2', null]);
mysql> select * from varchar_table;
+------+-------------------+
| key  | value             |
+------+-------------------+
|    1 | []                |
|    0 | NULL              |
|    2 | ['1', '2']        |
|    3 | [NULL, '2', NULL] |
+------+-------------------+
4 rows in set (0.01 sec)

# string
mysql> create table string_table (`key` int, value array<string>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into string_table values (0, null), (1, []), (2, ['1', '2']), (3, [null, '2', null]);
mysql> select * from string_table;
+------+-------------------+
| key  | value             |
+------+-------------------+
|    1 | []                |
|    0 | NULL              |
|    2 | ['1', '2']        |
|    3 | [NULL, '2', NULL] |
+------+-------------------+
4 rows in set (0.02 sec)

# decimal
mysql> create table decimal_table (`key` int, value array<decimal(27,9)>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into decimal_table values (0, null), (1, []), (2, [1.25, 2.25]), (3, [null, 2.25, null]);
mysql> select * from decimal_table;
+------+--------------------+
| key  | value              |
+------+--------------------+
|    0 | NULL               |
|    3 | [NULL, 2.25, NULL] |
|    2 | [1.25, 2.25]       |
|    1 | []                 |
+------+--------------------+
4 rows in set (0.02 sec)

# date
mysql> create table date_table (`key` int, value array<date>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into date_table values (0, null), (1, []), (2, ['2022-05-09', '2022-05-10']), (3, [null, '2022-05-10', null]);
mysql> select * from date_table;
+------+--------------------------+
| key  | value                    |
+------+--------------------------+
|    0 | NULL                     |
|    1 | []                       |
|    2 | [2022-05-09, 2022-05-10] |
|    3 | [NULL, 2022-05-10, NULL] |
+------+--------------------------+
4 rows in set (0.02 sec)

# datetime
mysql> create table datetime_table (`key` int, value array<datetime>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into datetime_table values (0, null), (1, []), (2, ['2022-05-09 14:00:00', '2022-05-10 14:30:00']), (3, [null, '2022-05-10 14:30:00', null]);
mysql> select * from datetime_table;
+------+--------------------------------------------+
| key  | value                                      |
+------+--------------------------------------------+
|    1 | []                                         |
|    0 | NULL                                       |
|    2 | [2022-05-09 14:00:00, 2022-05-10 14:30:00] |
|    3 | [NULL, 2022-05-10 14:30:00, NULL]          |
+------+--------------------------------------------+
4 rows in set (0.02 sec)

# nested array
mysql> create table nested_table (`key` int, value array<array<int>>) duplicate key (`key`) 
    -> distributed by hash (`key`) properties ('replication_num' = '1');
mysql> insert into nested_table values (0, null), (1, []), (2, [null]), (3, [[]]), (4, [[1, 2, 3]]), 
    -> (5, [null, [4, 5, 6], null]), (6, [[1, 2, 3], [4, 5, 6], [7, 8, 9]]);
mysql> select * from nested_table;
+------+-----------------------------------+
| key  | value                             |
+------+-----------------------------------+
|    4 | [[1, 2, 3]]                       |
|    5 | [NULL, [4, 5, 6], NULL]           |
|    1 | []                                |
|    0 | NULL                              |
|    2 | [NULL]                            |
|    3 | [[]]                              |
|    6 | [[1, 2, 3], [4, 5, 6], [7, 8, 9]] |
+------+-----------------------------------+
7 rows in set (0.02 sec)

```

### Related issues

#7570

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by adonis0147**
# Proposed changes

Issue Number: close #9465 

## Problem Summary:

Please refer to #9465 .

## Checklist(Required)

1. Does it affect the original behavior: No
2. Has unit tests been added: Yes
3. Has document been added or modified: No
4. Does it need to update dependencies: No
5. Are there any changes that cannot be rolled back: No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/core/block.cpp [hunk 1]
```diff
@@ -768,20 +768,38 @@ void Block::deep_copy_slot(void* dst, MemPool* pool, const doris::TypeDescriptor
...
+                                   item_offset, padding_char);
+                } else if (item_type_desc.is_date_type()) {
+                    // In CollectionValue, date type data is stored as either uint24_t or uint64_t.
+                    DateTimeValue datetime_value;
+                    deep_copy_slot(&datetime_value, pool, item_type_desc, data_ref, item_column,
+                                   item_offset, padding_char);
+                    DateTimeVal datetime_val;
+                    datetime_value.to_datetime_val(&datetime_val);
+                    iterator.set(&datetime_val);
+                } else {
```
Comment (cambyzju) [inline be/src/vec/core/block.cpp:792]: it is better use: if (is_data_type); else if (is_decimal_type); else (others);

File: be/src/util/array_parser.h [hunk 2]
```diff
@@ -175,6 +202,37 @@ class ArrayParser {
...
+            new (*val) DateTimeVal();
+            value.to_datetime_val(static_cast<DateTimeVal*>(*val));
+            break;
+        }
+        case TYPE_DECIMALV2: {
+            *val = reinterpret_cast<AnyVal*>(context->allocate(sizeof(DecimalV2Val)));
+            new (*val) DecimalV2Val();
+
+            if (iterator->IsNumber()) {
+                if (iterator->IsInt() || iterator->IsUint() || iterator->IsInt64()) {
```
Comment (cambyzju) [inline be/src/util/array_parser.h:219]: does IsInt32 better than IsInt, and IsUint32 better than IsUint?  At the first look, IsInt == IsInt32 || IsInt64.

File: be/src/util/array_parser.h [hunk 3]
```diff
@@ -175,6 +202,37 @@ class ArrayParser {
...
+                    DecimalV2Value(iterator->GetInt64(), 0)
+                            .to_decimal_val(static_cast<DecimalV2Val*>(*val));
+                } else if (iterator->IsUint64()) {
+                    DecimalV2Value(iterator->GetUint64(), 0)
```
Comment (cambyzju) [inline be/src/util/array_parser.h:223]: Why not use GetUint64 for IsUint?

File: be/src/runtime/collection_value.h [hunk 4]
```diff
@@ -137,45 +141,110 @@ struct CollectionValue {
...
+    bool has_next() const { return _offset < _collection_value->size(); }
+    bool next() const {
+        if (has_next()) {
+            ++_offset;
+            return true;
+        }
+        return false;
+    }
+    bool seek(int n) const {
+        if (n >= _collection_value->size()) {
```
Comment (cambyzju) [inline be/src/runtime/collection_value.h:158]: check n < 0?

Comment (adonis0147) [reply be/src/runtime/collection_value.h:159]: Fixed.

File: be/src/util/array_parser.h (hunk 3 above)
Comment (adonis0147) [reply be/src/util/array_parser.h:218]: Refined.

File: be/src/util/array_parser.h (hunk 2 above)
Comment (adonis0147) [reply be/src/util/array_parser.h]: After refinement, it is sufficient to check whether the iterator is `Uint64` here.

File: be/src/vec/olap/olap_data_convertor.cpp [hunk 5]
```diff
@@ -700,4 +621,55 @@ Status OlapBlockDataConvertor::OlapColumnDataConvertorDecimal::convert_to_olap()
...
+    assert(data_type_ptr_array);
+
+    CollectionValue* collection_value = _values.data();
+    for (int i = 0; i < _num_rows; ++i, ++collection_value) {
+        int64_t cur_pos = _row_pos + i;
+        int64_t prev_pos = cur_pos - 1;
+        if (_nullmap && _nullmap[cur_pos]) {
+            continue;
+        }
+        const auto& offsets = column_array->get_offsets();
```
Comment (cambyzju) [inline be/src/vec/olap/olap_data_convertor.cpp:649]: could put this line outside the for loop.

**Comment by cambyzju** [comment]
LGTM

File: be/src/vec/olap/olap_data_convertor.cpp (hunk 5 above)
Comment (adonis0147) [reply be/src/vec/olap/olap_data_convertor.cpp]: Refined.

**Comment by jackwener** [comment]
cc @platoneko

**Review by yangzhg — APPROVED** [review]
LGTM

File: be/src/vec/olap/olap_data_convertor.cpp [hunk 6]
```diff
@@ -700,4 +621,55 @@ Status OlapBlockDataConvertor::OlapColumnDataConvertorDecimal::convert_to_olap()
...
+            const auto& data_nullable_column =
+                    assert_cast<const ColumnNullable&>(column_array->get_data());
+            const auto* data_null_map = data_nullable_column.get_null_map_data().data();
+            collection_value->set_has_null(true);
+            collection_value->set_null_signs(
+                    const_cast<bool*>(reinterpret_cast<const bool*>(data_null_map + offset)));
+        }
+        ColumnWithTypeAndName item_typed_column = {column_array->get_data_ptr(),
+                                                   data_type_ptr_array->get_nested_type(), ""};
+        _item_convertor->set_source_column(item_typed_column, offset, size);
```
Comment (eldenmoon) [inline be/src/vec/olap/olap_data_convertor.cpp:668]: if item_type is or varchar or string, `set_source_column` will `resize()` the `_item_convertor._slice` and the slices from the previous `convert_to_olap()` will be freed, and then lead to heap use after free in the subsequent use of the collection_value

Comment (eldenmoon) [reply be/src/vec/olap/olap_data_convertor.cpp:668]: the same as array, hll or bitmap

Comment (adonis0147) [reply be/src/vec/olap/olap_data_convertor.cpp:668]: Hi @eldenmoon , you are right.

**Pull request merged**